Use canonical ordering in graphene

bissias opened the pull request:
This commit uses the canonical transaction order instead of sending ordering information provided that canonical ordering has been enabled and the graphene version is at least 1.

bissias committed:
working version with CTOR but no versioning yet

bissias committed:
add xversion to graphene

bissias committed:
ensure that rerequested transactions are placed in empty array locations

ptschip commented:
we're ready to start getting this in but needs a rebase

sickpig commented:
@bissias 

`ctor.py` seems to fail in a reproducible way on travis for linux64 and linux32 conf.

```
subprocess.TimeoutExpired: Command '['/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/build/../qa/rpc-tests/ctor.py', '--srcdir', '/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/build/src', '--coveragedir', '/tmp/coverageejbphnsl', '--no-ipv6-rpc-listen', '--portseed=175']' timed out after 30 seconds
```

bissias commented:
@sickpig thanks for the heads up. I just noticed that this seems to be failing on dev as well. I guess I should wait for it to get fixed there first.

sickpig commented:
@bissias I've tested locally current `dev` head (90f5ce45fcf9e096) and ctor.py is not failing:

```
2019-01-07 12:24:24,586.INFO: Random seed: 1546860264
Initializing test directory /tmp/test_ctorpy__m1z9yqs/28530
2019-01-07 12:24:25,600.INFO: Started bitcoind 0 as pid 28531 at 127.0.0.1:14645 dir /tmp/test_ctorpy__m1z9yqs/28530/node0
2019-01-07 12:24:26,615.INFO: Started bitcoind 1 as pid 28585 at 127.0.0.1:14646 dir /tmp/test_ctorpy__m1z9yqs/28530/node1
2019-01-07 12:24:27,631.INFO: Started bitcoind 2 as pid 28639 at 127.0.0.1:14647 dir /tmp/test_ctorpy__m1z9yqs/28530/node2
2019-01-07 12:24:28,645.INFO: Started bitcoind 3 as pid 28693 at 127.0.0.1:14648 dir /tmp/test_ctorpy__m1z9yqs/28530/node3
2019-01-07 12:24:29,519.INFO: sync blocks: [0, 0, 0, 0]
2019-01-07 12:24:29,969.INFO: sync blocks: [101, 33, 39, 3]
2019-01-07 12:24:31,006.INFO: sync blocks: [101, 101, 101, 54]
2019-01-07 12:24:32,014.INFO: sync blocks: [101, 101, 101, 101]
2019-01-07 12:24:33,032.INFO: Started bitcoind 6 as pid 29055 at 127.0.0.1:14651 dir /tmp/test_ctorpy__m1z9yqs/28530/node6
2019-01-07 12:24:35,184.INFO: sync blocks: [102, 101]
2019-01-07 12:24:36,189.INFO: sync blocks: [102, 102]
2019-01-07 12:24:36,226.INFO: sync blocks: [102, 101]
2019-01-07 12:24:37,230.INFO: sync blocks: [102, 102]
2019-01-07 12:24:37,267.INFO: sync blocks to 103: [103, 103]
2019-01-07 12:24:44,027.INFO: Started bitcoind 4 as pid 29262 at 127.0.0.1:14649 dir /tmp/test_ctorpy__m1z9yqs/28530/node4
2019-01-07 12:24:46,820.INFO: Started bitcoind 5 as pid 29458 at 127.0.0.1:14650 dir /tmp/test_ctorpy__m1z9yqs/28530/node5
2019-01-07 12:24:52,086.INFO: Stopping nodes
2019-01-07 12:24:52,293.INFO: Cleaning up
2019-01-07 12:24:52,299.INFO: Tests successful
```

I was able to reproduce the error while using your `ctor` branch though, see: 

```
2019-01-07 13:30:20,958.INFO: Random seed: 1546864220
Initializing test directory /tmp/test_ctorpy_stdmz807/19778
2019-01-07 13:30:21,972.INFO: Started bitcoind 0 as pid 19779 at 127.0.0.1:14503 dir /tmp/test_ctorpy_stdmz807/19778/node0
2019-01-07 13:30:22,987.INFO: Started bitcoind 1 as pid 19833 at 127.0.0.1:14504 dir /tmp/test_ctorpy_stdmz807/19778/node1
2019-01-07 13:30:24,002.INFO: Started bitcoind 2 as pid 19887 at 127.0.0.1:14505 dir /tmp/test_ctorpy_stdmz807/19778/node2
2019-01-07 13:30:25,019.INFO: Started bitcoind 3 as pid 19941 at 127.0.0.1:14506 dir /tmp/test_ctorpy_stdmz807/19778/node3
2019-01-07 13:30:25,892.INFO: sync blocks: [0, 0, 0, 0]
2019-01-07 13:30:26,148.INFO: sync blocks: [101, 9, 17, 2]
2019-01-07 13:30:27,162.INFO: sync blocks: [101, 101, 101, 43]
2019-01-07 13:30:28,170.INFO: sync blocks: [101, 101, 101, 101]
2019-01-07 13:30:29,187.INFO: Started bitcoind 6 as pid 20300 at 127.0.0.1:14509 dir /tmp/test_ctorpy_stdmz807/19778/node6
2019-01-07 13:30:31,370.INFO: sync blocks: [102, 102]
2019-01-07 13:30:31,374.ERROR: Unexpected exception caught during testing: StopIteration()
  File "/home/sickpig/src/BitcoinUnlimited/qa/rpc-tests/test_framework/test_framework.py", line 222, in main
    self.run_test()
  File "qa/rpc-tests/ctor.py", line 87, in run_test
    invalid = next(x for x in ct if x["status"] == "invalid")
2019-01-07 13:30:31,376.INFO: Stopping nodes
Traceback (most recent call last):
  File "qa/rpc-tests/ctor.py", line 215, in <module>
    MyTest ().main (bitcoinConfDict={"limitdescendantsize": 50,FORK_CFG: 0})
  File "/home/sickpig/src/BitcoinUnlimited/qa/rpc-tests/test_framework/test_framework.py", line 251, in main
    wait_bitcoinds()
  File "/home/sickpig/src/BitcoinUnlimited/qa/rpc-tests/test_framework/util.py", line 603, in wait_bitcoinds
    bitcoind.wait(timeout=BITCOIND_PROC_WAIT_TIMEOUT)
  File "/usr/lib/python3.5/subprocess.py", line 1650, in wait
    raise TimeoutExpired(self.args, timeout)
subprocess.TimeoutExpired: Command '['bitcoind', '-datadir=/tmp/test_ctorpy_stdmz807/19778/node6', '-rest', '-mocktime=0']' timed out after 60 seconds
```

bissias commented:
@sickpig you're right, sorry about that. I was pointing at the wrong binary. I'll look into the failure on my branch.

sickpig commented:
@bissias np at all :smile:

bissias committed:
don't use graphene in ctor.py

Graphene assumes that any sender with graphene xversion 1
will have CTOR enabled.

ptschip commented on src/blockrelay/graphene.cpp:
```diff
@@ -496,33 +511,56 @@ bool CGrapheneBlock::process(CNode *pfrom,
 
+        if (coinbase == NULL)
+            throw std::runtime_error("No coinbase transaction found in graphene block");
+
```
I think the coinbase check could be moved up to the same place where we do it in thinblocks... Take a look at IsThinBlockValid() in thinblock.cpp.

ptschip replied:
I suppose it depends on what your intent is though...are you expecting the graphene block to always have the coinbase?  In thinblocks we do, in compact blocks they don't force that but it eventually gets caught later because the block won't reconstruct...in graphene now, i can't remember if you MUST send the coinbase in the grapheneblock?

bissias replied:
For the purpose of CTOR it is essential that there exist a coinbase at this point and that I place it in the first position in the transaction vector. Otherwise, I cannot be certain that the ordering is correct in time for the Merkle root check.

ptschip replied:
One other note though, throwing exceptions can be an easy attack vector for filling up log files.  It would be better to have  a LOG(GRAPHENE, "your message...") statement with a false return...so it only gets logged if weve got logging on.

bissias replied:
Sure, I can make that change.

ptschip commented on src/blockrelay/graphene.cpp:
```diff
@@ -496,33 +511,56 @@ bool CGrapheneBlock::process(CNode *pfrom,
+
+                    if (it == blockCheapHashes.end())
+                        throw std::runtime_error("No coinbase transaction found in graphene block");
+
```
same thing here with throwing the exception.

bissias committed:
print to log and return false instead of throwing

ptschip commented on src/blockrelay/graphene.cpp:
```diff
@@ -559,7 +559,7 @@ bool CGrapheneBlock::process(CNode *pfrom,
 
                     if (it == blockCheapHashes.end())
-                        throw std::runtime_error("No coinbase transaction found in graphene block");
+                        return error("No coinbase transaction found in graphene block, peer=%s", pfrom->GetLogName());
```
this has the same issue...need to use:

LOG(GRAPHENE , "your mesage");
return false;

ptschip reviewed (approved):
Ran this overnight between two updated peers and seems to work fine.  Gets good compression when blocks get larger.  Saw one at 99.3% which have about 300 txns in it.
